src/main/kotlin/com/exactpro/th2/codec/EventProcessor.kt

```diff
@@ -30,91 +30,141 @@ import com.exactpro.th2.common.utils.message.sessionAlias
+            if (warnings.isNotEmpty()) {
+                val body = additionalBody()
+                warnings.forEach { warning ->
+                    onEachEvent(pairIds.toMap(), "[WARNING] During $action: $warning", body)
+                }
+            }
+        }
     }
 
-    fun onEachEvent(
-        parentEventIDs: Sequence<EventID>,
+    fun onErrorEvent(
+        eventId: EventID? = null,
         message: String,
         messagesIds: List<MessageID> = emptyList(),
-        body: List<String> = emptyList()
-    ) {
-        val warnEvent = codecEventID.onEvent(message, messagesIds, body)
-        parentEventIDs.forEach {
-            it.addReferenceTo(warnEvent, message, Event.Status.PASSED, messagesIds, additionalBody = body)
-        }
-    }
+        cause: Throwable? = null
+    ): EventID = (eventId ?: codecEventID).onErrorEvent(message, messagesIds, cause)
 
     fun onEachErrorEvent(
-        parentEventIDs: Sequence<EventID>,
+        pairIds: Map<MessageID, EventID?>,
         message: String,
-        messagesIds: List<MessageID> = emptyList(),
         cause: Throwable? = null,
-        additionalBody: List<String> = emptyList(),
-    ): EventID {
-        val errorEventId = codecEventID.onErrorEvent(message, messagesIds, cause, additionalBody)
-        parentEventIDs.forEach {
-            it.addReferenceTo(errorEventId, message, Event.Status.FAILED, messagesIds, cause, additionalBody)
+        body: List<String> = emptyList(),
+    ): Map<String, EventID> {
+        val bookToEventId = mutableMapOf<String, EventID>()
+        chooseRootEventsForPublication(pairIds).forEach { (codecEventId, externalEventIds) ->
+            val errorEventId = codecEventId.onErrorEvent(message, pairIds.keys, cause, body)
+            bookToEventId.put(errorEventId.bookName, errorEventId)?.also {
+                error(
+                    "Internal error: several root events have been chosen for pair ids: $pairIds"
+                )
+            }
+            externalEventIds.forEach {
+                it.addReferenceTo(errorEventId, message, Event.Status.FAILED, pairIds.keys, cause, body)
+            }
         }
-        return errorEventId
+        return bookToEventId
     }
 
-    fun onEachWarning(
-        parentEventIDs: Sequence<EventID>,
-        context: ReportingContext,
-        action: String,
-        additionalBody: () -> List<String> = ::emptyList,
-        messagesIds: () -> List<MessageID> = ::emptyList
-    ) = context.warnings.let { warnings ->
-        if (warnings.isNotEmpty()) {
-            val messages = messagesIds()
-            val body = additionalBody()
-            warnings.forEach { warning ->
-                onEachEvent(parentEventIDs, "[WARNING] During $action: $warning", messages, body)
+    internal fun chooseRootEventsForPublication(
+        pairIds: Map<MessageID, EventID?>,
+    ): Map<EventID, Set<EventID>> = mutableMapOf<EventID, MutableSet<EventID>>().apply {
```

Comment:
There is `buildMap` method in kotlin that can be used to fill the map. We can use that method here instead